File: ruby/Q66/Q66.rb
# David Lin
# 2017-08-02
# Decription: similar to adder in Electric Engineering
# Complexity: O(n)

# @param {Integer[]} digits
# @return {Integer[]}
def plus_one(digits)
  carry = 1 
  digits.reverse!.map! do |digit|
    digit = digit + carry
    carry = 0
    if digit == 10
      digit = 0
      carry = 1
    end
    digit
  end
  if carry == 1
    return [1] + digits.reverse # if the siginificant number is 9 and the array is not 0.
  else
    return digits.reverse
  end
end

if __FILE__ == $0 
  print plus_one([1,2,3,4])
  print plus_one([9])
  print plus_one([1,9])
end

File: javascript/Q66/Q66.js
/**
 * David Lin
 * 2018-05-28
 *
 * @param {number[]} digits
 * @return {number[]}
 */
var plusOne = function(digits) {
    digits = digits.reverse()
    var i = 0
    while (true) {
        if (digits[i] === undefined) {
            digits[i] = 0
        }
        digits[i] = (digits[i] + 1) % 10
        if (digits[i] !== 0) {
            return digits.reverse()
        }
        i = i + 1
    } 
};


